merkle-tree/concurrent/src/lib.rs

```diff
@@ -0,0 +1,360 @@
+                updated_proof[intersection_index] = changelog_entry.path[intersection_index];
+            } else {
+                return Err(HasherError::CannotUpdateLeaf);
+            }
+
+            k = (k + 1) % MAX_ROOTS;
+        }
+
+        Ok(updated_proof)
+    }
+
+    /// Updates the leaf under `leaf_index` with the `new_leaf` value.
+    ///
+    /// # Validation
+    ///
+    /// This method doesn't validate the proof. Caller is responsible for
+    /// doing that before.
+    fn update_leaf_in_tree(
+        &mut self,
+        new_leaf: &[u8; 32],
+        leaf_index: usize,
+        proof: &[[u8; 32]; MAX_HEIGHT],
+    ) -> Result<(), HasherError> {
+        let mut node = *new_leaf;
+        let mut changelog_path = [[0u8; 32]; MAX_HEIGHT];
+
+        for (j, sibling) in proof.iter().enumerate() {
+            changelog_path[j] = node;
+            node = compute_parent_node::<H>(&node, sibling, leaf_index, j)?;
+        }
+
+        let changelog_entry = ChangelogEntry::new(node, changelog_path, leaf_index);
+        self.inc_current_changelog_index();
+        if let Some(changelog_element) = self
+            .changelog
+            .get_mut(self.current_changelog_index as usize)
+        {
+            *changelog_element = changelog_entry
+        }
+
+        self.inc_current_root_index();
+        *self
+            .roots
+            .get_mut(self.current_root_index as usize)
+            .ok_or(HasherError::RootsZero)? = node;
+
+        if self.rightmost_index > 0 && leaf_index == self.rightmost_index as usize - 1 {
+            self.rightmost_proof.copy_from_slice(proof);
+            self.rightmost_leaf = *new_leaf;
+        }
+
+        Ok(())
+    }
+
+    /// Replaces the `old_leaf` under the `leaf_index` with a `new_leaf`, using
+    /// the given `proof` and `changelog_index` (pointing to the changelog entry
+    /// which was the newest at the time of preparing the proof).
+    pub fn update(
+        &mut self,
+        changelog_index: usize,
+        old_leaf: &[u8; 32],
+        new_leaf: &[u8; 32],
+        leaf_index: usize,
+        proof: &[[u8; 32]],
+    ) -> Result<(), HasherError> {
+        // Provided proof might have lower number of elements than the height.
+        // In such case, fill the proof with zero bytes.
+        let proof: [[u8; 32]; MAX_HEIGHT] = if proof.len() < MAX_HEIGHT {
+            full_proof::<H, MAX_HEIGHT>(proof)
+        } else {
+            proof.try_into().map_err(|_| HasherError::ProofTooLarge)?
+        };
+
+        let updated_proof = self.update_proof(changelog_index, leaf_index, &proof)?;
+
+        if is_valid_proof::<H, MAX_HEIGHT>(
+            &self.roots[self.current_root_index as usize],
+            old_leaf,
+            leaf_index,
+            &proof,
+        )? {
+            self.update_leaf_in_tree(new_leaf, leaf_index, &updated_proof)?
+        }
+
+        Ok(())
```

Comment:
```suggestion
        is_valid_proof::<H, MAX_HEIGHT>(
            &self.roots[self.current_root_index as usize],
            old_leaf,
            leaf_index,
            &proof,
        )?;
        self.update_leaf_in_tree(new_leaf, leaf_index, &updated_proof)
```

I think this would make the code less noisy.

Reply:
An alternative could be to move the is valid proof call into update proof.